File: examples/ruby/escrow/Escrow.runar.rb
require 'runar'

class Escrow < Runar::SmartContract
  prop :buyer, PubKey
  prop :seller, PubKey
  prop :arbiter, PubKey

  def initialize(buyer, seller, arbiter)
    super(buyer, seller, arbiter)
    @buyer = buyer
    @seller = seller
    @arbiter = arbiter
  end

  runar_public seller_sig: Sig, arbiter_sig: Sig
  def release(seller_sig, arbiter_sig)
    assert check_sig(seller_sig, @seller)
    assert check_sig(arbiter_sig, @arbiter)
  end

  runar_public buyer_sig: Sig, arbiter_sig: Sig
  def refund(buyer_sig, arbiter_sig)
    assert check_sig(buyer_sig, @buyer)
    assert check_sig(arbiter_sig, @arbiter)
  end
end

File: conformance/tests/escrow/escrow.runar.rs
use runar::prelude::*;

#[runar::contract]
struct Escrow {
    #[readonly]
    buyer: PubKey,
    #[readonly]
    seller: PubKey,
    #[readonly]
    arbiter: PubKey,
}

#[runar::methods(Escrow)]
impl Escrow {
    #[public]
    fn release(&self, seller_sig: Sig, arbiter_sig: Sig) {
        assert!(check_sig(seller_sig, self.seller));
        assert!(check_sig(arbiter_sig, self.arbiter));
    }

    #[public]
    fn refund(&self, buyer_sig: Sig, arbiter_sig: Sig) {
        assert!(check_sig(buyer_sig, self.buyer));
        assert!(check_sig(arbiter_sig, self.arbiter));
    }
}
